lib/src/source/git.dart

```diff
@@ -253,9 +302,22 @@ class GitSource extends CachedSource {
   /// by [id] on the effective ref of [id].
   ///
   /// This assumes that the canonical clone already exists.
-  Future<String> _getRev(PackageId id) {
-    return git.run(["rev-list", "--max-count=1", _getEffectiveRef(id)],
-        workingDir: _repoCachePath(id)).then((result) => result.first);
+  Future<String> _getRev(PackageId id) async {
+    var ref = _getEffectiveRef(id);
```

Comment:
I'm having an issue with this. Here's what the code roughly looks like:

```dart
Future<String> _getRev(PackageId id) {
  print('_getRev called: ${id.toString()}');
  ...

  var ref = description is Map ? description['ref'] : null;
  if (ref == null && id.version != null && id.version != Version.none) {
    print('Trying to use version: ${id.version.toString()}');
   // return the rev-list $version result
  } else {
    print('Returning _getRef(id) or HEAD');
    // return rev-list _getRef(id) ?? 'HEAD';
  }
}
```

For some reason `id.version` ends up getting set to something even if no `version` is specified in the pubspec. So if I put some print statements in this `_getRev` method the output looks like this when I run a simple test like `test/get/git/check_out_test.dart`

```
  | MSG : Resolving dependencies...
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 1.0.0 from git
  | Trying to use version: 1.0.0
```
That's with a description of `foo: "../foo.git"` (no version specified). So by the logic in the if condition, it will try to get a revision with `1.0.0` which fails as it should. It should have hit the `else` case instead for this to work properly.

If I leverage `_useVersionTags(id)` in that if condition I can correctly determine whether or not the version should be used as a ref. This is what the existing code in `_getEffectiveRef` is doing. However, you mentioned you don't want to rely on an additional annotation for this to work.

I will keep looking for a solution, but let me know if you have any ideas that will solve this. Then I should be able to get rid of the `_useVersionTags(id)` stuff.